File: services/delivery/src/main/java/io/wkrzywiec/fooddelivery/delivery/infra/stream/RedisStreamConfiguration.java
package io.wkrzywiec.fooddelivery.delivery.infra.stream;

import com.fasterxml.jackson.databind.ObjectMapper;
import io.wkrzywiec.fooddelivery.commons.infra.messaging.MessagePublisher;
import io.wkrzywiec.fooddelivery.commons.infra.messaging.redis.RedisMessageConsumerConfig;
import io.wkrzywiec.fooddelivery.commons.infra.messaging.redis.RedisStreamPublisher;
import io.wkrzywiec.fooddelivery.delivery.domain.DeliveryFacade;
import lombok.extern.slf4j.Slf4j;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;
import org.springframework.context.annotation.Profile;
import org.springframework.data.redis.connection.RedisConnectionFactory;
import org.springframework.data.redis.core.RedisTemplate;
import org.springframework.data.redis.serializer.StringRedisSerializer;
import org.springframework.data.redis.stream.Subscription;

@Slf4j
@Configuration
@Profile("redis-stream")
public class RedisStreamConfiguration extends RedisMessageConsumerConfig {

    @Bean
    public Subscription ordersChannelSubscription(RedisConnectionFactory factory,
                                                  RedisTemplate<String, String> redisTemplate,
                                                  RedisOrdersChannelConsumer streamListener) {
        return createSubscription(redisTemplate, factory, streamListener);
    }

    @Bean
    public RedisOrdersChannelConsumer redisOrdersChannelConsumer(DeliveryFacade facade, ObjectMapper objectMapper) {
        return new RedisOrdersChannelConsumer(facade, objectMapper);
    }

    @Bean
    public MessagePublisher messagePublisher(RedisTemplate<String, String> redisTemplate, ObjectMapper objectMapper) {
        return new RedisStreamPublisher(redisTemplate, objectMapper);
    }

    @Bean
    public RedisTemplate<String, ?> redisTemplate(RedisConnectionFactory connectionFactory) {
        RedisTemplate<String, Object> redisTemplate = new RedisTemplate<String, Object>();
        redisTemplate.setConnectionFactory(connectionFactory);
        redisTemplate.setKeySerializer(new StringRedisSerializer());
        redisTemplate.setHashKeySerializer(new StringRedisSerializer());
        redisTemplate.setValueSerializer(new StringRedisSerializer());
        redisTemplate.setHashValueSerializer(new StringRedisSerializer());
        redisTemplate.afterPropertiesSet();

        return redisTemplate;
    }
}

File: services/ordering/src/main/kotlin/io/wkrzywiec/fooddelivery/ordering/infra/stream/RedisStreamConfiguration.kt
package io.wkrzywiec.fooddelivery.ordering.infra.stream

import com.fasterxml.jackson.databind.ObjectMapper
import io.wkrzywiec.fooddelivery.commons.infra.messaging.MessagePublisher
import io.wkrzywiec.fooddelivery.commons.infra.messaging.redis.RedisMessageConsumerConfig
import io.wkrzywiec.fooddelivery.commons.infra.messaging.redis.RedisStreamListener
import io.wkrzywiec.fooddelivery.commons.infra.messaging.redis.RedisStreamPublisher
import io.wkrzywiec.fooddelivery.ordering.domain.OrderingFacade
import lombok.extern.slf4j.Slf4j
import org.springframework.context.annotation.Bean
import org.springframework.context.annotation.Configuration
import org.springframework.context.annotation.Profile
import org.springframework.data.redis.connection.RedisConnectionFactory
import org.springframework.data.redis.core.RedisTemplate
import org.springframework.data.redis.serializer.StringRedisSerializer
import org.springframework.data.redis.stream.Subscription

@Slf4j
@Configuration
@Profile("redis-stream")
class RedisStreamConfiguration : RedisMessageConsumerConfig() {
    @Bean
    fun ordersChannelSubscription(
        factory: RedisConnectionFactory?,
        redisTemplate: RedisTemplate<String?, String?>?,
        streamListener: RedisStreamListener?
    ): Subscription {
        return createSubscription(redisTemplate, factory, streamListener)
    }

    @Bean
    fun redisOrdersChannelConsumer(facade: OrderingFacade, objectMapper: ObjectMapper): RedisStreamListener {
        return RedisOrdersChannelConsumer(facade, objectMapper)
    }

    @Bean
    fun messagePublisher(
        redisTemplate: RedisTemplate<String?, String?>?,
        objectMapper: ObjectMapper?
    ): MessagePublisher {
        return RedisStreamPublisher(redisTemplate, objectMapper)
    }

    @Bean
    fun redisTemplate(connectionFactory: RedisConnectionFactory?): RedisTemplate<String, *> {
        val redisTemplate = RedisTemplate<String, Any>()
        redisTemplate.connectionFactory = connectionFactory
        redisTemplate.keySerializer = StringRedisSerializer()
        redisTemplate.hashKeySerializer = StringRedisSerializer()
        redisTemplate.valueSerializer = StringRedisSerializer()
        redisTemplate.hashValueSerializer = StringRedisSerializer()
        redisTemplate.afterPropertiesSet()

        return redisTemplate
    }
}
